packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js

```diff
@@ -0,0 +1,113 @@
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {AnyNativeEvent} from '../PluginModuleType';
+import type {DOMEventName} from '../DOMEventNames';
+import type {DispatchQueue} from '../DOMPluginEventSystem';
+import type {EventSystemFlags} from '../EventSystemFlags';
+import type {Fiber} from 'react-reconciler/src/ReactInternalTypes';
+
+import {getFiberCurrentPropsFromNode} from '../../client/ReactDOMComponentTree';
+
+import {SyntheticEvent} from '../SyntheticEvent';
+
+/**
+ * This plugin invokes action functions on forms, inputs and buttons if
+ * the form doesn't prevent default.
+ */
+function extractEvents(
+  dispatchQueue: DispatchQueue,
+  domEventName: DOMEventName,
+  targetInst: null | Fiber,
+  nativeEvent: AnyNativeEvent,
+  nativeEventTarget: null | EventTarget,
+  eventSystemFlags: EventSystemFlags,
+  targetContainer: EventTarget,
+) {
+  if (domEventName !== 'submit') {
+    return;
+  }
+  if (!targetInst || targetInst.stateNode !== nativeEventTarget) {
+    // If we're inside a parent root that itself is a parent of this root, then
+    // its deepest target won't be the actual form that's being submitted.
+    return;
+  }
+  const form: HTMLFormElement = (nativeEventTarget: any);
+  let action = (getFiberCurrentPropsFromNode(form): any).action;
+  const submitter: null | HTMLInputElement | HTMLButtonElement =
+    (nativeEvent: any).submitter;
+  let submitterAction;
+  if (submitter) {
+    const submitterProps = getFiberCurrentPropsFromNode(submitter);
+    submitterAction = submitterProps
+      ? (submitterProps: any).formAction
+      : submitter.getAttribute('formAction');
+    if (submitterAction != null) {
+      // The submitter overrides the form action.
+      action = submitterAction;
+    }
+  }
+
+  if (typeof action !== 'function') {
+    return;
+  }
+
+  const event = new SyntheticEvent(
+    'action',
+    'action',
+    null,
+    nativeEvent,
+    nativeEventTarget,
+  );
+
+  function submitForm() {
+    if (nativeEvent.defaultPrevented) {
+      // We let earlier events to prevent the action from submitting.
+      return;
+    }
+    // Prevent native navigation.
+    event.preventDefault();
+    let formData;
+    if (submitter) {
+      // The submitter's value should be included in the FormData.
+      // It should be in the document order in the form.
+      // Since the FormData constructor invokes the formdata event it also
+      // needs to be available before that happens so after construction it's too
+      // late. The easiest way to do this is to switch the form field to hidden,
+      // which is always included, and then back again. This does means that this
+      // is observable from the formdata event though.
+      // TODO: This tricky doesn't work on button elements. Consider inserting
+      // a fake node instead for that case.
+      // TODO: FormData takes a second argument that it's the submitter but this
```

Comment:
This is available in the latest versions of all major browsers now: https://developer.mozilla.org/en-US/docs/Web/API/FormData/FormData#browser_compatibility ... should be available in jest+jsdom [very soon](https://github.com/jsdom/jsdom/pull/3496) too 🤞

Reply:
Unfortunately not all browsers are truly evergreen. We support a lot of old Samsung TVs and many iOS users don't upgrade for years. So we can't use it yet. Not worth the added code to do the image one.

Reply:
Note that we do it a little differently now. https://github.com/facebook/react/blob/main/packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js#L92-L97

Reply:
That makes sense, thanks for the added context 👍 

Though if you're supporting old browsers, many don't support `SubmitEvent` so you won't have a `submitter` here in the first place. And for modern browsers, a typical `submitter` won't work with the `type="hidden"` hack (i.e. IME `<button>` is much more prevalent than `<input type=submit>`).

Would it be better to just remove the hack altogether and pass the `submitter` (if any) to `FormData`? Then you'd get the right behavior in the common case, with less code.

Reply:
We don't support very old, like we don't support IE anymore, however we support somewhat old browser like TVs and iOS versions a year or so old. Point taken that maybe it's not even enough to rely on the `submitter` property on the event.

I'd expect us to only keep this hack around for maybe 6 months or so until we switch to the native version given this awkward transition stage.